### ISSUE TITLE
Syncthing "Receive Encrypted" mode tries to create filenames too long for ext4 filesystem

### ISSUE THREAD

**Issue by DeflateAwning** [syncthing/syncthing#9441]
When I setup Syncthing in "Receive Encrypted" mode, Syncthing running on the computer receiving the encrypted files tries to create files with filenames too long for the ext4 filesystem which is storing the files.

While this is partly a limitation of ext4, it would be nice if Syncthing was still able to find a workaround to avoid trying to create these files which are too long, as I don't really have a choice of another filesystem.

* Syncthing version: v1.27.3, Linux (64-bit Intel/AMD), "Gold Grasshopper", Build 2024-01-14 (noupgrade)
* Operating system: Linux Mint 21.3
* Browser and version: N/A

**Comment by calmh** [comment]
ext4 supports path components up to 255 characters, Syncthing uses up to 200. What error are you actually seeing?

**Comment by DeflateAwning** [comment]
This is the error message from the logs. It gets repeated in the logs a lot:

```
2024-03-02 18:10:55 Puller (folder "1xxxxxxx" (7xxxx-12345), item "N.syncthing-enc/JK/G5B18RMND3J7CE5OJHKIFMAFKLDMI9LS81NKQ66FKRG8K34JP0K1Q5G9JIS58519S85IJLV5EI1SHM6QJ647QDA3DET1BS85JN9L45E2SCD1C538HD5LO09E1VV5KV0O8SPH44EUTT3TA9VDQL22JLNT18VNM92NUMTE028"): syncing: finishing: checking existing file: lstat /home/xxxx/1xxxxxxx/N.syncthing-enc/JK/G5B18RMND3J7CE5OJHKIFMAFKLDMI9LS81NKQ66FKRG8K34JP0K1Q5G9JIS58519S85IJLV5EI1SHM6QJ647QDA3DET1BS85JN9L45E2SCD1C538HD5LO09E1VV5KV0O8SPH44EUTT3TA9VDQL22JLNT18VNM92NUMTE028: file name too long
2024-03-02 18:10:55 "1xxxxxxx" (7xxxx-12345): Failed to sync 587 items
2024-03-02 18:10:55 Folder "1xxxxxxx" (7xxxx-12345) isn't making sync progress - retrying in 1h6m0s.
```

Folder names, paths, etc. have been redacted. The exact lengths in characters remain though (i.e., the length of `x` characters is the exact same length as the original names).

**Comment by DeflateAwning** [comment]
Upon further investigation, this appears to be related to #3338. My home directory is actually an EncFS filesystem, I think.

[The EncFS Wikipedia](https://en.wikipedia.org/wiki/EncFS) suggests that the max filename length is 190 bytes:

> Most filesystems limit filenames to 255 bytes; in that case, EncFS only supports filenames up to 190 bytes.

Is there any chance that Syncthing would consider decreasing the max generated filename even further?

**Comment by calmh** [comment]
Yeah, that's unfortunate. We considered the limitations we were aware of, but encfs was not one of those things. There are a couple of other design warts with how the filename encoding works that should also be fixed, but doing so requires changing how they're represented in the wire protocol and migrating the files on disk etc and is a bit cumbersome so it's not an easy fix.

(Of course, your computer so you do what you want to, but it seems to me like encrypted shares and encfs do essentially the same things and layering them may not buy you much extra security. One might also have hoped that the likes of encfs had abstracted this away and not exposed the encryption overhead as an extra limitation.)

**Comment by DeflateAwning** [comment]
An example case where the double-encryption is beneficial is if you're using a work computer as a backup node (i.e., the boss/IT has the drive encryption keys), but you still want to back you syncthing files to it. This is roughly my case. As such, I'll leave this here as a request.

**Comment by we-sell-bags** [comment]
> Upon further investigation, this appears to be related to #3338. My home directory is actually an EncFS filesystem, I think.
> 
> [The EncFS Wikipedia](https://en.wikipedia.org/wiki/EncFS) suggests that the max filename length is 190 bytes:
> 
> > Most filesystems limit filenames to 255 bytes; in that case, EncFS only supports filenames up to 190 bytes.
> 
> Is there any chance that Syncthing would consider decreasing the max generated filename even further?


it's potentially worse than that.....
you seem to be  assuming the encoding is standard and not something like MBC Chinese or other Asian Characters, so potentially your path/ filename can far exceed the max for most file systems.
normally any Path name in non "English". had to be divided by 3 or 4., so if it looks like it is 10 char long, it can actually be 30 or 40  long... fortunately  the filenames are machine generated, but the paths are user.
so simply checking the file name length is not enough... because you are syncing to paths.

so whilst the file name may meet the "maximum" requirements allowed , it may  well still break the overall constraints  once the path is included...even though it's not breaking the "file name" constraints.

**Comment by calmh** [comment]
That has nothing to do with Syncthing encrypted path components, which are always maximum 200 ASCII characters.

**Comment by imsodin** [comment]
Just for completeness: The limit is for a go string length, which is bytes. So even if it was non-ASCII with multi-byte chars here, the limit would still be correct:

> len(s)    string type      string length in bytes

https://go.dev/ref/spec#Length_and_capacity

**Comment by daisaacson** [comment]
I'm trying to sync to a Synology NAS and I'm receiving the same `file name too long` error. The Synology NAS folder is encrypted, it looks like under the hood Synology is using `ecryptfs`. It was their solution for encrypting data at rest back in DSM 6.x. There is a Note on the encrypted folder:

> The name of a file or a folder within the encrypted share folder cannot exceed 143 English characters or 47 Asian (CJK) characters.

I'm testing a `syncthing` setup prior to running this on a friend's Synology NAS where I do want to have `syncthing` `Receive Encrypted` enabled.

**Comment by daisaacson** [comment]
I think what I'm going to attempt to do, after looking at https://github.com/syncthing/syncthing/issues/3338 and a165838 is under Actions > Advanced > Folders > Folder 'xyz', add an additional variable that overrides `maxFilenameLength` for the edge cases like mine. Seems like the current default value works for most.

**Comment by daisaacson** [comment]
@calmh, you are too fast.
My thought was to do an Advanced feature that could override the `maxFilenameLengh` per folder, for edge case like ours.
My vision would be to have an additional field available
![Image](https://github.com/user-attachments/assets/87e8408a-4283-43e4-ad14-b344b95a51a7)

I couldn't get past a container error

> does not match regexp ^v\d+\.\d+\.\d+(-[a-z0-9]+)*(\.\d+)*(\+\d+-g[0-9a-f]+|\+[0-9a-z]+)?(-[^\s]+)?$ 
> 2025/06/22 15:12:42 Invalid version string "c63032e4-dirty-fix-9941"; 

I will cease my efforts, since you are miles ahead of me.
Thanks for all your work.
